### PR TITLE
Reimplement string.Replace in managed code

### PR DESCRIPTION
**Pull request by jkotas**
Prep-work for moving String.Manipulation.cs to shared partition. Avoiding manually managed code overhead makes it quite a bit faster as well in some cases. For example:

`"hello".Replace("a", "b")` - 1.5x faster
`"hello".Replace("e", "a")` - 1.05x faster
`"1000000x a".Replace("a", "b")` - 1.5x faster

### CODE REVIEW

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 1]
```diff
@@ -1079,18 +1087,97 @@ public String Replace(char oldChar, char newChar)
...
+            unsafe
+            {
+                fixed (char* pThis = &_firstChar)
+                {
+                    int matchIdx = 0;
+                    int lastPossibleMatchIdx = this.Length - oldValue.Length;
+                    while (matchIdx <= lastPossibleMatchIdx)
+                    {
+                        char* pMatch = pThis + matchIdx;
+                        for (int probeIdx = 0; probeIdx < oldValue.Length; probeIdx++)
```
Comment (jkotas) [inline src/mscorlib/src/System/String.Manipulation.cs]: This is potential candidate for SequenceEquals once it becomes available in CoreLib.

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 2]
```diff
@@ -1079,18 +1087,97 @@ public String Replace(char oldChar, char newChar)
...
-            // Note that if newValue is null, we treat it like String.Empty.
+            if (oldValue.Length == 0)
+                throw new ArgumentException(SR.Format(SR.Argument_StringZeroLength, nameof(oldValue)));
+
+            // Api behavior: if newValue is null, instances of oldValue are to be removed.
+            if (newValue == null)
+                newValue = string.Empty;
+
+            // Workaround for https://github.com/dotnet/coreclr/issues/16197
+            // Span<int> initialSpan = stackalloc int[StackallocIntBufferSizeLimit];
```
Comment (danmosemsft) [inline src/mscorlib/src/System/String.Manipulation.cs]: Have we considered having a constructor on ValueListBuilder that takes care of stackallocing some initial span possibly with a hint as to the size? That would save callers repeating that work.

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 3]
```diff
@@ -1079,18 +1087,97 @@ public String Replace(char oldChar, char newChar)
...
+                int count = replacementIdx - thisIdx;
+                if (count != 0)
+                {
+                    thisSpan.Slice(thisIdx, count).CopyTo(dstSpan.Slice(dstIdx));
+                    dstIdx += count;
+                }
+                thisIdx = replacementIdx + oldValue.Length;
+
+                // Copy over newValue to replace the oldValue.
+                newValue.AsReadOnlySpan().CopyTo(dstSpan.Slice(dstIdx));
```
Comment (danmosemsft) [inline src/mscorlib/src/System/String.Manipulation.cs]: why not pull out `newValue.AsReadOnlySpan()` into a local like `thisSpan` ? You need it repeatedly.

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/src/System/String.Manipulation.cs]: You cannot do that unfortunately. The stackallocated memory is gone when the method returns. So if you stackalloc inside ValueListBuilder constructor, the memory is gone when the constructor returns.

Comment (danmosemsft) [reply src/mscorlib/src/System/String.Manipulation.cs]: Ah of course.

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 3 above)
Comment (jkotas) [reply src/mscorlib/src/System/String.Manipulation.cs]: It actually hurts to cache the Span in a local like that. The code became a bit/smaller faster when I have removed caching of `thisSpan` as well.

The problem is that Span is (pointer+length). It burns two registers and causes you to run out of registers quickly. It is better to recompute it each time here.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 4]
```diff
@@ -1079,18 +1087,96 @@ public String Replace(char oldChar, char newChar)
...
+                                goto Next;
+                            }
+                        }
+                        // Found a match for the string. Record the location of the match and skip over the "oldValue."
+                        replacementIndices.Append(matchIdx);
+                        matchIdx += oldValue.Length;
+                        continue;
+
+                    Next:
+                        matchIdx++;
```
Comment (danmosemsft) [inline src/mscorlib/src/System/String.Manipulation.cs]: In future we could consider adding code to drop to Boyer Moore if both replacement string and text are large.

File: src/vm/object.h [hunk 5]
```diff
@@ -2678,7 +2678,6 @@ class StringBufferObject : public Object
 
     static void ReplaceBuffer(STRINGBUFFERREF *thisRef, __in_ecount(newLength) WCHAR *newBuffer, INT32 newLength);
     static void ReplaceBufferAnsi(STRINGBUFFERREF *thisRef, __in_ecount(newCapacity) CHAR *newBuffer, INT32 newCapacity);
```
Comment (danmosemsft) [inline src/vm/object.h]: Are these guys dead too? I can't see a caller to ReplaceBufferAnsi only ReplaceBufferAnsiInternal. Unless some macro magic measn they're the same.

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 6]
```diff
@@ -1079,18 +1087,96 @@ public String Replace(char oldChar, char newChar)
...
+            if (oldValue.Length == 0)
+                throw new ArgumentException(SR.Argument_StringZeroLength, nameof(oldValue));
+
+            // Api behavior: if newValue is null, instances of oldValue are to be removed.
+            if (newValue == null)
+                newValue = string.Empty;
+
+            // Workaround for https://github.com/dotnet/coreclr/issues/16197
+            // Span<int> initialSpan = stackalloc int[StackallocIntBufferSizeLimit];
+            Span<int> initialSpan; StackallocIntBuffer initialBuffer; unsafe { initialSpan = new Span<int>(&initialBuffer, StackallocIntBufferSizeLimit); }
```
Comment (danmosemsft) [inline src/mscorlib/src/System/String.Manipulation.cs]: Do you want to do the same temporary hack further down, replacing `Span<int> initialSpan = stackalloc int[StackallocIntBufferSizeLimit];` for the Split ?

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 7]
```diff
@@ -1079,18 +1087,96 @@ public String Replace(char oldChar, char newChar)
...
-
-        public String Replace(String oldValue, String newValue)
+        public string Replace(string oldValue, string newValue)
         {
             if (oldValue == null)
                 throw new ArgumentNullException(nameof(oldValue));
-            // Note that if newValue is null, we treat it like String.Empty.
+            if (oldValue.Length == 0)
+                throw new ArgumentException(SR.Argument_StringZeroLength, nameof(oldValue));
+
```
Comment (danmosemsft) [inline src/mscorlib/src/System/String.Manipulation.cs]: bail out early if oldValue same as newValue - at least Reference.equals anyway?

Strictly a breaking change as you no longer get a new instance.

File: src/mscorlib/src/System/String.Manipulation.cs [hunk 8]
```diff
@@ -1079,18 +1087,96 @@ public String Replace(char oldChar, char newChar)
...
+            string dst = ReplaceHelper(oldValue.Length, newValue, replacementIndices.AsReadOnlySpan());
+
+            replacementIndices.Dispose();
+
+            return dst;
+        }
+
+        private string ReplaceHelper(int oldValueLength, string newValue, ReadOnlySpan<int> indices)
+        {
+            long dstLength = this.Length + ((long)(newValue.Length - oldValueLength)) * indices.Length;
```
Comment (danmosemsft) [inline src/mscorlib/src/System/String.Manipulation.cs]: Debug.Assert(indices.Length > 0) ?

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 3 above)
Comment (danmosemsft) [reply src/mscorlib/src/System/String.Manipulation.cs]: Ah intersting.

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 6 above)
Comment (jkotas) [reply src/mscorlib/src/System/String.Manipulation.cs]: Those other methods do not suffer from the codegen bug as much. They do not have loops in them.

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 4 above)
Comment (jkotas) [reply src/mscorlib/src/System/String.Manipulation.cs]: https://github.com/dotnet/coreclr/issues/6918

**Comment by danmosemsft** [comment]
Why is this faster? Overhead of the single native call? The old implementation allocating a indices buffer the size of the whole target string?

**Comment by jkotas** [comment]
> Why is this faster?

Every line in the C++ implementation that looks unnatural and does not have direct equivalent in the C# implementation is extra overhead of manually managed code: `HELPER_METHOD_FRAME_BEGIN_RET_PROTECT` macro is expensive. Reporting all object references to the GC manually is not free either.

File: src/vm/object.h (hunk 5 above)
Comment (jkotas) [reply src/vm/object.h]: They are called from here: https://github.com/dotnet/coreclr/blob/master/src/classlibnative/bcltype/stringbuffer.cpp#L29

Comment (danmosemsft) [reply src/vm/object.h]: Uh, I forgot that there is a world outside of /vm/ that uses this header.

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 7 above)
Comment (jkotas) [reply src/mscorlib/src/System/String.Manipulation.cs]: It is not clear whether this optimization would kick in frequently enough to pay for itself. We have a number of discussions about similar special cases in various PRs.

File: src/mscorlib/src/System/String.Manipulation.cs (hunk 8 above)
Comment (jkotas) [reply src/mscorlib/src/System/String.Manipulation.cs]: Fixed

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline

**Pull request merged**